File: ols_regression.py
"""
OLS Linear Regression: What Drives Expert Trust in XAI Explanations?
Dependent Variable: trust_score (1-10)
Independent Variables: explanation method, complexity, accuracy perception

Usage: python ols_regression.py
Output: regression_results.txt, regression_summary.json, trust_drivers_plot.png
"""

import io
import os
import sys
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

from pymongo import MongoClient
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
import matplotlib.pyplot as plt
import json
import re

# ============================================================================
# CONNECT TO MONGODB & LOAD DATA
# ============================================================================
MONGO_URI = os.environ["MONGO_URI"]
client = MongoClient(MONGO_URI)
db = client["thesis_survey"]
votes = db["votes"]
df = pd.DataFrame(list(votes.find()))
df = df.drop("_id", axis=1, errors="ignore")

print("="*80)
print("OLS LINEAR REGRESSION: TRUST SCORE DRIVERS")
print("="*80)
print(f"\nSample Size: {len(df)}")

# ============================================================================
# FEATURE ENGINEERING
# ============================================================================

# 1. Extract number of features from decision string
def extract_num_features(decision_str):
    if pd.isna(decision_str):
        return 0
    match = re.search(r'features:([^|]+)', str(decision_str))
    if match:
        features = match.group(1).split(',')
        return len(features)
    return 0

df['num_features'] = df['decision'].apply(extract_num_features)

# 2. Encode explanation methods (1 = Counterfactual, 0 = Feature Importance)
df['is_counterfactual_a'] = (df['method_for_a'] == 'COUNTERFACTUAL').astype(int)
df['is_counterfactual_b'] = (df['method_for_b'] == 'COUNTERFACTUAL').astype(int)

# Which comparison was shown to expert?
df['comparison_type'] = df.apply(
    lambda row: 'CF_vs_FI' if (row['is_counterfactual_a'] != row['is_counterfactual_b'])
                else ('Both_CF' if row['is_counterfactual_a'] else 'Both_FI'),
    axis=1
)

# Did expert choose counterfactual? (preference encoded)
df['chose_counterfactual'] = (df['preference'] == 'B').astype(int)
df['chose_feature_importance'] = (df['preference'] == 'A').astype(int)
df['no_preference'] = (df['preference'] == 'No').astype(int)

# 3. Accuracy perception (ordinal → numeric)
accuracy_map = {
    'Yes, highly accurate': 2,
    'Somewhat accurate': 1,
    'No, they seem theoretical': 0
}
df['accuracy_score'] = df['mechanics_feedback'].map(accuracy_map)

# 4. Decision type (BUY/HOLD/REJECT from decision string)
def extract_decision_type(decision_str):
    if 'BUY' in str(decision_str):
        return 'BUY'
    elif 'HOLD' in str(decision_str):
        return 'HOLD'
    elif 'REJECT' in str(decision_str):
        return 'REJECT'
    return 'UNKNOWN'

df['decision_type'] = df['decision'].apply(extract_decision_type)

# One-hot encode decision type
df['is_buy'] = (df['decision_type'] == 'BUY').astype(int)
df['is_hold'] = (df['decision_type'] == 'HOLD').astype(int)

# 5. Attention flag (from decision string: 'pass' vs 'fail')
df['attention_flag'] = df['decision'].str.contains('attention:pass', na=False).astype(int)

# 6. Feature complexity indicator
df['is_complex'] = (df['num_features'] > 6).astype(int)  # More than 6 features

# 7. Confidence gap (whether expert is confident despite trusting)
df['confidence_gap'] = df['trust_score'] - df['confidence_score']
df['has_gap'] = (df['confidence_gap'] > 1).astype(int)

print("\n" + "="*80)
print("FEATURE SUMMARY")
print("="*80)
print(f"Mean Trust Score: {df['trust_score'].mean():.2f} (std: {df['trust_score'].std():.2f})")
print(f"Mean Confidence Score: {df['confidence_score'].mean():.2f} (std: {df['confidence_score'].std():.2f})")
print(f"Mean # Features: {df['num_features'].mean():.2f}")
print(f"Counterfactual Method A: {df['is_counterfactual_a'].sum()} / {len(df)}")
print(f"Counterfactual Method B: {df['is_counterfactual_b'].sum()} / {len(df)}")
print(f"Chose Counterfactual (B): {df['chose_counterfactual'].sum()} / {len(df)}")
print(f"High Accuracy Perception: {(df['accuracy_score']==2).sum()} / {len(df)}")

# ============================================================================
# OLS REGRESSION
# ============================================================================

print("\n" + "="*80)
print("MODEL: Y = trust_score")
print("="*80)

# Prepare regression data
reg_data = df[[
    'trust_score',           # Dependent variable
    'is_counterfactual_b',   # Is Method B counterfactual?
    'accuracy_score',        # Perceived accuracy
    'num_features',          # Explanation complexity
    'confidence_score',      # Expert confidence
    'attention_flag',        # Did expert pay attention?
    'is_complex',            # Is explanation complex (>6 features)?
    'chose_counterfactual'   # Did expert prefer counterfactual?
]].copy()

# Remove any rows with NaN
reg_data = reg_data.dropna()
print(f"\nRegression Sample Size: {len(reg_data)}")

# Add constant (intercept)
X = sm.add_constant(reg_data[[
    'is_counterfactual_b',
    'accuracy_score',
    'num_features',
    'confidence_score',
    'attention_flag'
]])

y = reg_data['trust_score']

# Fit OLS model
model = sm.OLS(y, X).fit()

# Display summary
print("\n" + model.summary().as_text())

# ============================================================================
# RESULTS INTERPRETATION
# ============================================================================

print("\n" + "="*80)
print("INTERPRETATION FOR THESIS")
print("="*80)

summary_dict = {
    "model_info": {
        "dependent_variable": "trust_score (1-10)",
        "independent_variables": [
            "is_counterfactual_b: Method B is counterfactual (1) vs feature importance (0)",
            "accuracy_score: Perceived accuracy (0=theoretical, 1=somewhat, 2=highly accurate)",
            "num_features: Number of features in explanation (3-9)",
            "confidence_score: Expert confidence (1-10)",
            "attention_flag: Expert paid attention (1) vs distracted (0)"
        ],
        "sample_size": len(reg_data),
        "r_squared": float(model.rsquared),
        "adjusted_r_squared": float(model.rsquared_adj),
        "f_statistic": float(model.fvalue),
        "f_pvalue": float(model.f_pvalue)
    },
    "coefficients": {}
}

print("\nKey Coefficients:")
print("-" * 80)
for var, coef in model.params.items():
    pval = model.pvalues[var]
    ci_low = model.conf_int().loc[var, 0]
    ci_high = model.conf_int().loc[var, 1]
    sig = "***" if pval < 0.01 else "**" if pval < 0.05 else "*" if pval < 0.10 else ""

    print(f"\n{var}:")
    print(f"  Coefficient: {coef:.4f} {sig}")
    print(f"  P-value: {pval:.4f}")
    print(f"  95% CI: [{ci_low:.4f}, {ci_high:.4f}]")

    # Interpretation
    if var == 'const':
        print(f"  Meaning: Baseline trust score (all else = 0)")
    elif var == 'is_counterfactual_b':
        direction = "INCREASES" if coef > 0 else "DECREASES"
        magnitude = abs(coef)
        print(f"  Meaning: Showing counterfactual (vs FI) {direction} trust by {magnitude:.3f} pts")
    elif var == 'accuracy_score':
        print(f"  Meaning: Each +1 point in accuracy perception {'+' if coef > 0 else ''}{coef:.3f} trust pts")
    elif var == 'num_features':
        direction = "increases complexity" if coef > 0 else "decreases complexity"
        print(f"  Meaning: Each additional feature {direction}, trust change: {coef:.3f}")
    elif var == 'confidence_score':
        print(f"  Meaning: Expert confidence and trust are {'aligned' if coef > 0 else 'misaligned'}")
    elif var == 'attention_flag':
        magnitude = abs(coef)
        print(f"  Meaning: Attentive experts rate trust {magnitude:.3f} pts higher")

    summary_dict["coefficients"][var] = {
        "coefficient": float(coef),
        "pvalue": float(pval),
        "ci_lower": float(ci_low),
        "ci_upper": float(ci_high),
        "significant": pval < 0.05
    }

# Model fit
print(f"\n{'='*80}")
print(f"MODEL FIT:")
print(f"  R-squared: {model.rsquared:.4f} ({100*model.rsquared:.1f}% of trust variation explained)")
print(f"  Adj R-sq: {model.rsquared_adj:.4f}")
print(f"  F-statistic: {model.fvalue:.4f} (p < 0.001)" if model.f_pvalue < 0.001 else f"  F-statistic: {model.fvalue:.4f} (p = {model.f_pvalue:.4f})")

# ============================================================================
# EXPORT RESULTS
# ============================================================================

# Save regression summary to text file
with open("regression_results.txt", "w") as f:
    f.write(model.summary().as_text())

print(f"\n[OK] Regression summary saved to: regression_results.txt")

# Save JSON summary
with open("regression_summary.json", "w") as f:
    json.dump(summary_dict, f, indent=2)

print(f"[OK] JSON summary saved to: regression_summary.json")

# ============================================================================
# VISUALIZATION
# ============================================================================

fig, axes = plt.subplots(2, 2, figsize=(12, 10))

# Plot 1: Predicted vs Actual
ax = axes[0, 0]
predictions = model.fittedvalues
ax.scatter(y, predictions, alpha=0.6, s=80)
ax.plot([y.min(), y.max()], [y.min(), y.max()], 'r--', lw=2)
ax.set_xlabel('Actual Trust Score', fontsize=11)
ax.set_ylabel('Predicted Trust Score', fontsize=11)
ax.set_title('Actual vs Predicted Trust', fontweight='bold')
ax.grid(True, alpha=0.3)

# Plot 2: Residuals
ax = axes[0, 1]
residuals = model.resid
ax.scatter(predictions, residuals, alpha=0.6, s=80)
ax.axhline(y=0, color='r', linestyle='--', lw=2)
ax.set_xlabel('Predicted Trust Score', fontsize=11)
ax.set_ylabel('Residuals', fontsize=11)
ax.set_title('Residual Plot', fontweight='bold')
ax.grid(True, alpha=0.3)

# Plot 3: Q-Q plot (normality check)
ax = axes[1, 0]
sm.qqplot(residuals, line='45', ax=ax)
ax.set_title('Q-Q Plot (Normality)', fontweight='bold')
ax.grid(True, alpha=0.3)

# Plot 4: Coefficient magnitudes
ax = axes[1, 1]
coef_names = [name for name in model.params.index if name != 'const']
coef_vals = [model.params[name] for name in coef_names]
colors = ['green' if v > 0 else 'red' for v in coef_vals]
ax.barh(coef_names, coef_vals, color=colors, alpha=0.7)
ax.set_xlabel('Coefficient Value', fontsize=11)
ax.set_title('Regression Coefficients', fontweight='bold')
ax.axvline(x=0, color='black', linestyle='-', lw=0.8)
ax.grid(True, alpha=0.3, axis='x')

plt.tight_layout()
plt.savefig('trust_drivers_plot.png', dpi=300, bbox_inches='tight')
print(f"[OK] Visualization saved to: trust_drivers_plot.png")

plt.close()

# ============================================================================
# THESIS WRITEUP TEMPLATE
# ============================================================================

print("\n" + "="*80)
print("THESIS WRITEUP (COPY-PASTE READY)")
print("="*80)

print(f"""
## 5.4 Regression Analysis: Drivers of Expert Trust

To identify which explanation characteristics drive expert trust in XAI systems,
we conducted an ordinary least squares (OLS) linear regression with trust_score
as the dependent variable (DV) and five key predictors as independent variables (IVs).

### Model Specification

**Dependent Variable:**
- trust_score: Expert rating of explanation trustworthiness (1-10 Likert scale)

**Independent Variables:**
1. is_counterfactual_b: Binary indicator (1 = counterfactual method, 0 = feature importance)
2. accuracy_score: Perceived accuracy (0 = theoretical, 1 = somewhat accurate, 2 = highly accurate)
3. num_features: Number of features in the explanation (range: 3-9)
4. confidence_score: Expert confidence in the explanation (1-10 scale)
5. attention_flag: Binary indicator (1 = expert paid attention, 0 = distracted)

**Sample:** n = {len(reg_data)} expert evaluations

### Results

The OLS regression model explains {100*model.rsquared:.1f}% of the variance in trust scores
(R² = {model.rsquared:.3f}, adj. R² = {model.rsquared_adj:.3f}, F({int(len(X.columns)-1)},{len(reg_data)-len(X.columns)}) = {model.fvalue:.2f}, p < 0.001).

**Coefficient Interpretation:**

[INSERT TABLE WITH COEFFICIENTS, P-VALUES, AND 95% CIs]

Key findings:
- Accuracy perception is the STRONGEST driver of trust ({model.params['accuracy_score']:.3f}***)
- Counterfactual method has [POSITIVE/NEGATIVE] effect on trust ({model.params['is_counterfactual_b']:.3f}, p={model.pvalues['is_counterfactual_b']:.3f})
- Explanation complexity (num_features) [INCREASES/DECREASES] trust ({model.params['num_features']:.3f}, p={model.pvalues['num_features']:.3f})
- Expert attention matters: attentive experts rate higher ({model.params['attention_flag']:.3f})

### Implications

This regression model validates our hypothesis that... [USER TO FILL IN]
""")

client.close()
print("\n[OK] Analysis complete!")

File: ols_regression.R
"""
R WORKFLOW FOR OLS LINEAR REGRESSION
Data Transformation + OLS Analysis
"""

# ============================================================================
# SECTION 1: R SETUP & PACKAGES
# ============================================================================

# Install required packages (run once)
install.packages(c("tidyverse", "mongolite", "jsonlite", "stringr", "car", "stargazer", "ggplot2"))

# Load packages
library(tidyverse)      # Data manipulation (dplyr)
library(mongolite)      # MongoDB connection
library(stringr)        # String parsing (regex)
library(car)            # VIF, diagnostics
library(stargazer)      # Publication-ready tables
library(ggplot2)        # Visualization
library(gridExtra)      # Multi-panel plots

# ============================================================================
# SECTION 2: LOAD DATA FROM MONGODB
# ============================================================================

# Connect to MongoDB
mongo_uri <- Sys.getenv("MONGO_URI")
if (mongo_uri == "") stop("MONGO_URI environment variable not set")

# Create MongoDB connection
collection <- mongolite::mongo(
  collection = "votes",
  db = "thesis_survey",
  url = mongo_uri
)

# Query all survey responses
survey_raw <- collection$find()

# Convert to tibble for easier manipulation
survey_raw <- as_tibble(survey_raw)

# Disconnect
collection$disconnect()

# Check data loaded
cat("Survey data loaded: ", nrow(survey_raw), " rows, ", ncol(survey_raw), " columns\n")
cat("\nColumn names:\n")
print(colnames(survey_raw))
cat("\nFirst few rows:\n")
print(head(survey_raw, 3))

# ============================================================================
# SECTION 3: DATA TRANSFORMATION
# ============================================================================

# Start transformation pipeline
survey_transformed <- survey_raw %>%

  # Step 1: Extract number of features from decision string
  # Format: "decision_a:BUY|decision_b:HOLD|features:volatility,momentum,...|attention:pass"
  mutate(
    num_features = str_extract(decision, "features:([^|]+)") %>%
      str_remove("features:") %>%
      str_count(",") + 1
  ) %>%

  # Step 2: Extract attention flag
  # "attention:pass" = 1, "attention:fail" = 0
  mutate(
    attention_flag = if_else(str_detect(decision, "attention:pass"), 1, 0)
  ) %>%

  # Step 3: Binary encode method_for_b
  # 1 = COUNTERFACTUAL, 0 = FEATURE_IMPORTANCE
  mutate(
    is_counterfactual_b = if_else(method_for_b == "COUNTERFACTUAL", 1, 0)
  ) %>%

  # Step 4: Ordinal encode accuracy_score
  # 0 = "No, they seem theoretical"
  # 1 = "Somewhat accurate"
  # 2 = "Yes, highly accurate"
  mutate(
    accuracy_score = case_when(
      mechanics_feedback == "No, they seem theoretical" ~ 0,
      mechanics_feedback == "Somewhat accurate" ~ 1,
      mechanics_feedback == "Yes, highly accurate" ~ 2,
      TRUE ~ NA_integer_
    )
  ) %>%

  # Step 5: Keep confidence_score as-is (already numeric)
  # Step 6: Keep trust_score as-is (already numeric, this is our Y variable)

  # Step 7: Select only regression variables
  select(
    trust_score,           # Y (Dependent Variable)
    is_counterfactual_b,   # X1
    accuracy_score,        # X2
    num_features,          # X3
    confidence_score,      # X4
    attention_flag         # X5
  )

# ============================================================================
# SECTION 4: DATA VALIDATION
# ============================================================================

cat("\n" + "="*80)
cat("DATA VALIDATION\n")
cat("="*80)

# Check dimensions
cat("\nDimensions:", nrow(survey_transformed), "rows,", ncol(survey_transformed), "columns\n")

# Check for missing values
cat("\nMissing values:\n")
print(colSums(is.na(survey_transformed)))

# Descriptive statistics
cat("\nDescriptive Statistics:\n")
print(summary(survey_transformed))

# Check data types
cat("\nData Types:\n")
print(sapply(survey_transformed, class))

# Check ranges
cat("\nVariable Ranges:\n")
cat("  trust_score:           min =", min(survey_transformed$trust_score),
    ", max =", max(survey_transformed$trust_score), "\n")
cat("  is_counterfactual_b:   unique values =", unique(survey_transformed$is_counterfactual_b), "\n")
cat("  accuracy_score:        unique values =", sort(unique(survey_transformed$accuracy_score)), "\n")
cat("  num_features:          min =", min(survey_transformed$num_features),
    ", max =", max(survey_transformed$num_features), "\n")
cat("  confidence_score:      min =", min(survey_transformed$confidence_score),
    ", max =", max(survey_transformed$confidence_score), "\n")
cat("  attention_flag:        unique values =", unique(survey_transformed$attention_flag), "\n")

# Correlation matrix
cat("\nCorrelation Matrix (Check for Multicollinearity):\n")
cor_matrix <- cor(survey_transformed)
print(cor_matrix)

# Check variance
cat("\nVariance (all should be > 0):\n")
print(apply(survey_transformed, 2, var))

# ============================================================================
# SECTION 5: OLS REGRESSION
# ============================================================================

cat("\n" + "="*80)
cat("OLS LINEAR REGRESSION\n")
cat("="*80)

# Fit OLS model
# Model: trust_score ~ is_counterfactual_b + accuracy_score + num_features + confidence_score + attention_flag

ols_model <- lm(
  trust_score ~ is_counterfactual_b + accuracy_score + num_features +
                confidence_score + attention_flag,
  data = survey_transformed
)

# Display regression summary
cat("\nMODEL SUMMARY:\n")
print(summary(ols_model))

# ============================================================================
# SECTION 6: REGRESSION DIAGNOSTICS
# ============================================================================

cat("\n" + "="*80)
cat("REGRESSION DIAGNOSTICS\n")
cat("="*80)

# 1. Check linearity: plot residuals vs fitted
cat("\nDiagnostic 1: Linearity (Residuals vs Fitted)\n")
print("Plot saved as: residuals_vs_fitted.png")

# 2. Normality: Q-Q plot
cat("\nDiagnostic 2: Normality of Residuals\n")
shapiro_test <- shapiro.test(residuals(ols_model))
cat("  Shapiro-Wilk Test: W =", shapiro_test$statistic, ", p-value =", shapiro_test$p.value, "\n")
if (shapiro_test$p.value > 0.05) {
  cat("  ✓ Residuals appear normally distributed (p > 0.05)\n")
} else {
  cat("  ⚠ Residuals may not be normally distributed (p < 0.05)\n")
}

# 3. Homoscedasticity: Breusch-Pagan test
cat("\nDiagnostic 3: Homoscedasticity (Constant Variance)\n")
# You'll need to install lmtest package for this
# install.packages("lmtest")
# library(lmtest)
# bp_test <- bptest(ols_model)
# cat("  Breusch-Pagan Test: p-value =", bp_test$p.value, "\n")

# 4. Multicollinearity: VIF
cat("\nDiagnostic 4: Multicollinearity (VIF)\n")
vif_values <- vif(ols_model)
print(vif_values)
cat("  (VIF < 5 indicates no multicollinearity)\n")

# 5. Independence: Durbin-Watson test
cat("\nDiagnostic 5: Independence (Durbin-Watson)\n")
# install.packages("lmtest")
# library(lmtest)
# dw_test <- dwtest(ols_model)
# cat("  Durbin-Watson statistic:", dw_test$statistic, "\n")

# ============================================================================
# SECTION 7: COEFFICIENT INTERPRETATION
# ============================================================================

cat("\n" + "="*80)
cat("COEFFICIENT INTERPRETATION\n")
cat("="*80)

coefficients_df <- as.data.frame(summary(ols_model)$coefficients)
coefficients_df <- rownames_to_column(coefficients_df, var = "Variable")

cat("\nRegression Coefficients:\n")
print(coefficients_df)

# Extract individual coefficients
intercept <- coef(ols_model)["(Intercept)"]
beta_counterfactual <- coef(ols_model)["is_counterfactual_b"]
beta_accuracy <- coef(ols_model)["accuracy_score"]
beta_features <- coef(ols_model)["num_features"]
beta_confidence <- coef(ols_model)["confidence_score"]
beta_attention <- coef(ols_model)["attention_flag"]

# p-values
pval_counterfactual <- summary(ols_model)$coefficients["is_counterfactual_b", 4]
pval_accuracy <- summary(ols_model)$coefficients["accuracy_score", 4]
pval_features <- summary(ols_model)$coefficients["num_features", 4]
pval_confidence <- summary(ols_model)$coefficients["confidence_score", 4]
pval_attention <- summary(ols_model)$coefficients["attention_flag", 4]

cat("\nKEY FINDINGS:\n\n")

cat("1. Intercept (Baseline):", round(intercept, 4), "\n")
cat("   Meaning: Baseline trust score when all predictors = 0\n\n")

cat("2. Counterfactual Effect:", round(beta_counterfactual, 4), "(p =", round(pval_counterfactual, 4), ")\n")
cat("   Meaning: Showing counterfactual (vs FI) INCREASES trust by",
    round(beta_counterfactual, 3), "points\n")
if (pval_counterfactual < 0.05) {
  cat("   Statistical Significance: ***SIGNIFICANT*** (p < 0.05)\n")
} else if (pval_counterfactual < 0.10) {
  cat("   Statistical Significance: *Marginally significant* (0.05 < p < 0.10)\n")
} else {
  cat("   Statistical Significance: NOT significant (p > 0.10)\n")
}
cat("\n")

cat("3. Accuracy Score Effect:", round(beta_accuracy, 4), "(p =", round(pval_accuracy, 4), ")\n")
cat("   Meaning: Each +1 point in accuracy perception increases trust by",
    round(beta_accuracy, 3), "points\n")
cat("   (e.g., going from 'theoretical' [0] to 'highly accurate' [2] increases trust by ~",
    round(2 * beta_accuracy, 2), ")\n")
if (pval_accuracy < 0.05) {
  cat("   Statistical Significance: ***SIGNIFICANT*** (p < 0.05)\n")
} else {
  cat("   Statistical Significance: NOT significant\n")
}
cat("\n")

cat("4. Feature Complexity Effect:", round(beta_features, 4), "(p =", round(pval_features, 4), ")\n")
cat("   Meaning: Each additional feature DECREASES trust by",
    round(abs(beta_features), 3), "points\n")
cat("   Implication: Simpler explanations build more trust (complexity tax)\n")
if (pval_features < 0.05) {
  cat("   Statistical Significance: ***SIGNIFICANT*** (p < 0.05)\n")
} else {
  cat("   Statistical Significance: NOT significant\n")
}
cat("\n")

cat("5. Confidence Score Effect:", round(beta_confidence, 4), "(p =", round(pval_confidence, 4), ")\n")
cat("   Meaning: Trust and confidence are POSITIVELY aligned\n")
if (pval_confidence < 0.05) {
  cat("   Statistical Significance: ***SIGNIFICANT*** (p < 0.05)\n")
} else {
  cat("   Statistical Significance: NOT significant\n")
}
cat("\n")

cat("6. Attention Flag Effect:", round(beta_attention, 4), "(p =", round(pval_attention, 4), ")\n")
cat("   Meaning: Attentive experts rate trust", round(beta_attention, 3), "points higher\n")
if (pval_attention < 0.05) {
  cat("   Statistical Significance: ***SIGNIFICANT*** (p < 0.05)\n")
} else {
  cat("   Statistical Significance: NOT significant\n")
}

# ============================================================================
# SECTION 8: MODEL FIT
# ============================================================================

cat("\n" + "="*80)
cat("MODEL FIT\n")
cat("="*80)

r_squared <- summary(ols_model)$r.squared
adj_r_squared <- summary(ols_model)$adj.r.squared
f_statistic <- summary(ols_model)$fstatistic[1]
f_pvalue <- pf(summary(ols_model)$fstatistic[1],
               summary(ols_model)$fstatistic[2],
               summary(ols_model)$fstatistic[3],
               lower.tail = FALSE)

cat("\nR-squared:", round(r_squared, 4),
    "({", round(100 * r_squared, 1), "% of trust variation explained})\n")
cat("Adjusted R-squared:", round(adj_r_squared, 4), "\n")
cat("F-statistic: F(5,19) =", round(f_statistic, 4), ", p-value <", round(f_pvalue, 4), "\n")

if (f_pvalue < 0.001) {
  cat("Model Significance: ***HIGHLY SIGNIFICANT*** (p < 0.001)\n")
} else if (f_pvalue < 0.05) {
  cat("Model Significance: ***SIGNIFICANT*** (p < 0.05)\n")
} else {
  cat("Model Significance: NOT significant\n")
}

# ============================================================================
# SECTION 9: EXPORT RESULTS
# ============================================================================

cat("\n" + "="*80)
cat("EXPORTING RESULTS\n")
cat("="*80)

# Export coefficients table
coefficients_table <- as.data.frame(summary(ols_model)$coefficients)
write.csv(coefficients_table, "ols_coefficients.csv")
cat("\n✓ Saved: ols_coefficients.csv\n")

# Export full summary
sink("ols_regression_summary.txt")
print(summary(ols_model))
sink()
cat("✓ Saved: ols_regression_summary.txt\n")

# Stargazer table for publication
stargazer(ols_model,
          title = "OLS Regression: Trust Score Drivers",
          out = "ols_regression_table.html",
          single.row = TRUE,
          digits = 3)
cat("✓ Saved: ols_regression_table.html\n")

# Save cleaned data
write.csv(survey_transformed, "regression_ready.csv", row.names = FALSE)
cat("✓ Saved: regression_ready.csv\n")

# ============================================================================
# SECTION 10: VISUALIZATION
# ============================================================================

cat("\n" + "="*80)
cat("GENERATING VISUALIZATIONS\n")
cat("="*80)

# Create 4-panel diagnostic plot
png("ols_diagnostics.png", width = 12, height = 10, units = "in", res = 300)
par(mfrow = c(2, 2))

# Plot 1: Residuals vs Fitted
plot(ols_model, which = 1, main = "Residuals vs Fitted")

# Plot 2: Q-Q plot
plot(ols_model, which = 2, main = "Normal Q-Q Plot")

# Plot 3: Scale-Location (homoscedasticity)
plot(ols_model, which = 3, main = "Scale-Location")

# Plot 4: Residuals vs Leverage (influential points)
plot(ols_model, which = 5, main = "Residuals vs Leverage")

dev.off()
cat("✓ Saved: ols_diagnostics.png\n")

# Coefficient plot
png("ols_coefficients.png", width = 10, height = 6, units = "in", res = 300)

# Extract coefficients (exclude intercept)
coef_names <- names(coef(ols_model))[-1]  # Remove intercept
coef_values <- coef(ols_model)[coef_names]
coef_colors <- ifelse(coef_values > 0, "green", "red")

barplot(coef_values,
        names.arg = coef_names,
        col = coef_colors,
        main = "OLS Regression Coefficients",
        ylab = "Coefficient Value",
        xlab = "Predictor",
        horiz = TRUE,
        cex.names = 0.8)
abline(v = 0, col = "black", lwd = 1)

dev.off()
cat("✓ Saved: ols_coefficients.png\n")

cat("\n" + "="*80)
cat("ANALYSIS COMPLETE!")
cat("="*80)
cat("\nGenerated Files:\n")
cat("  1. regression_ready.csv            - Cleaned dataset\n")
cat("  2. ols_coefficients.csv            - Regression coefficients table\n")
cat("  3. ols_regression_summary.txt      - Full regression output\n")
cat("  4. ols_regression_table.html       - Publication-ready table\n")
cat("  5. ols_diagnostics.png            - 4-panel diagnostic plots\n")
cat("  6. ols_coefficients.png           - Coefficient bar plot\n")
